### ISSUE TITLE
HIGH: [Polish] [$500] Finalize order of LHN in #focus mode

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#33029]
**Problem:**
When the LHN is in "focus" mode, its sorting should be:

1. All pinned reports, alphabetically
2. Everything else, alphabetically

However, it doesn't seem to be.  Here's from @iwiznia:

![Image](https://github.com/Expensify/App/assets/575186/eff71010-258b-465e-b02c-14efe6f52327)

**Solution:**
Review the LHN logic for sorting in focus mode to understand what it's currently doing, and verify it's correct.  The correct behavior should be:

*#focus mode:*
* All pinned chats (alphabetically)
* Everything else (alphabetically)

*most recent mode:*
* All pinned chats (alphabetically)
* Everything else (most recent sorting)

Additionally:

* When sorting alphabetically, it should only sort the alphanumeric characters -- other characters should be ignored.  
* "pinned chats" refers to chats that are actually pinned, as well as any chat with a "green/red brick road" (GBR/RBR), which are treated as if they are pinned.

So, for this issue please:

1. Review the [existing sorting code](https://github.com/Expensify/App/blob/adf68dcb0f1b649dfcc32d075789c57bffd4a5cb/src/libs/SidebarUtils.ts#L164-L172)
2. Update it to match the logic above
3. Update the sort comparator to ignore non-alphanumeric characters when sorting
4. Verify that GBR chats are showing as pinned
5. Verify that RBR chats are showing as pinned

You up to the challenge?  Let's see those proposals!!
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01432274cc2b58438c</li>
        <li>Upwork Job ID: 1744194942807576576</li>
        <li>Last Price Increase: 2024-01-08</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @youssef-lr</details>

**Comment by quinthar** [comment]
*Problem:*
When the LHN is in "focus" mode, its sorting should be:

1. All pinned reports, alphabetically
2. Everything else, alphabetically

However, it doesn't seem to be.  Here's from @iwiznia:

![Image](https://github.com/Expensify/App/assets/575186/eff71010-258b-465e-b02c-14efe6f52327)

*Solution:*
Review the LHN logic for sorting in focus mode to understand what it's currently doing, and verify it's correct.  In the above screenshot it _could_ be that we are sorting `#` ahead of `E` -- maybe we should sort only on the alphanumeric characters?

**Comment by quinthar** [comment]
Upgrading to HIGH as it's fundamentally about making it work correctly.

**Comment by shubham1206agra** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue

Finalize order of LHN in #focus mode

### What is the root cause of that problem?

This should be fix with `getOrderedReportIDs`

### What changes do you think we should make in order to solve the problem?

We should modify the `getOrderedReportIDs` logic to the above set of rules to get the desired result.

### What alternative solutions did you explore? (Optional)

**Comment by thebigshotsam** [comment]
# LHN Sorting Logic Proposal

## Problem Statement
When the LHN (Left Hand Navigation) is in "focus" mode, the existing sorting logic was not aligning with the expected behavior. The sorting criteria were not correctly implemented, specifically for pinned reports and those in "focus" mode.

## Solution
The LHN sorting logic has been reviewed and updated to meet the following requirements:

### Sorting Criteria
1. **Focus Mode:**
   - All pinned chats should be sorted alphabetically.
   - Everything else should be sorted alphabetically.

2. **Most Recent Mode:**
   - All pinned chats should be sorted alphabetically.
   - Everything else should be sorted based on the most recent activity.

### Additional Considerations
- When sorting alphabetically, non-alphanumeric characters are ignored.
- "Pinned chats" include both actually pinned chats and those with a GBR (Green Brick Road) or RBR (Red Brick Road).
- GBR chats and RBR chats should be treated as pinned.

## Implementation Details

### Sorting Function
The existing sorting function has been modified to achieve the desired behavior. Here are the key updates:

1. **Grouping:**
   - Reports are grouped into four categories: Pinned/GBR, Drafts, Non-archived reports, and Archived reports.

2. **Sorting:**
   - Pinned/GBR reports and Drafts are sorted alphabetically.
   - Non-archived reports are sorted based on the most recent activity in "Most Recent" mode and alphabetically in "Focus" mode.
   - Archived reports are sorted similarly to non-archived reports based on the mode.

3. **Combining Arrays:**
   - Pinned/GBR reports, Drafts, Non-archived reports, and Archived reports are combined into a single array (`allReportsSorted`).

4. **Final Sorting:**
   - The combined array is sorted based on the specified criteria.

5. **Extracting ReportIDs:**
   - The final array is used to extract reportIDs, and the cache is updated accordingly.

## Usage
The `sortLHNReports` function should be used to obtain the sorted list of reportIDs for the LHN.

```typescript
const sortedLHNReports = sortLHNReports(reportsToDisplay, isInDefaultMode);
// Function to sort LHN reports
function sortLHNReports(reportsToDisplay: Report[], isInDefaultMode: boolean): string[] {
    const pinnedAndGBRReports: Report[] = [];
    const draftReports: Report[] = [];
    const nonArchivedReports: Report[] = [];
    const archivedReports: Report[] = [];

    // Calculate properties and categorize reports
    reportsToDisplay.forEach((report) => {
        report.displayName = ReportUtils.getReportName(report);
        report.iouReportAmount = ReportUtils.getMoneyRequestReimbursableTotal(report, allReports);

        const isPinned = report.isPinned ?? false;
        const reportAction = ReportActionsUtils.getReportAction(report.parentReportID ?? '', report.parentReportActionID ?? '');

        if (isPinned || ReportUtils.requiresAttentionFromCurrentUser(report, reportAction)) {
            pinnedAndGBRReports.push(report);
        } else if (report.hasDraft) {
            draftReports.push(report);
        } else if (ReportUtils.isArchivedRoom(report)) {
            archivedReports.push(report);
        } else {
            nonArchivedReports.push(report);
        }
    });

    // Sort each group of reports accordingly
    pinnedAndGBRReports.sort((a, b) => (a.displayName && b.displayName ? a.displayName.localeCompare(b.displayName) : 0));
    draftReports.sort((a, b) => (a.displayName && b.displayName ? a.displayName.localeCompare(b.displayName) : 0));

    if (isInDefaultMode) {
        // Sort non-archived reports based on most recent activity and alphabetically
        nonArchivedReports.sort((a, b) => {
            const compareDates = a.lastVisibleActionCreated && b.lastVisibleActionCreated ? compareStringDates(b.lastVisibleActionCreated, a.lastVisibleActionCreated) : 0;
            const compareDisplayNames = a.displayName && b.displayName ? a.displayName.localeCompare(b.displayName) : 0;
            return compareDates || compareDisplayNames;
        });

        // For archived reports, ensure most recent reports are at the top by reversing the order
        archivedReports.sort((a, b) => (a.lastVisibleActionCreated && b.lastVisibleActionCreated ? compareStringDates(b.lastVisibleActionCreated, a.lastVisibleActionCreated) : 0));
    } else {
        // Sort non-archived and archived reports alphabetically
        nonArchivedReports.sort((a, b) => (a.displayName && b.displayName ? a.displayName.localeCompare(b.displayName) : 0));
        archivedReports.sort((a, b) => (a.displayName && b.displayName ? a.displayName.localeCompare(b.displayName) : 0));
    }

    // Combine all groups into a single array for final sorting
    const allReportsSorted: Report[] = [...pinnedAndGBRReports, ...draftReports, ...nonArchivedReports, ...archivedReports];

    // Sort the combined array based on specified criteria
    allReportsSorted.sort((a, b) => {
        const comparePinned = (b.isPinned ? 1 : 0) - (a.isPinned ? 1 : 0);
        const compareDisplayNames = a.displayName && b.displayName ? a.displayName.localeCompare(b.displayName) : 0;
        const compareDates = a.lastVisibleActionCreated && b.lastVisibleActionCreated ? compareStringDates(b.lastVisibleActionCreated, a.lastVisibleActionCreated) : 0;

        // First, sort by pinned status, then by display names, and finally by dates in most recent mode
        return comparePinned || compareDisplayNames || compareDates;
    });

    // Extract reportIDs from the final sorted array
    const LHNReports = allReportsSorted.map((report) => report.reportID);

    // Update the cache with the sorted reportIDs
    setWithLimit(reportIDsCache, cachedReportsKey, LHNReports);

    return LHNReports;
}

// Function to compare string dates
function compareStringDates(dateStringA: string, dateStringB: string): number {
    const dateA = new Date(dateStringA);
    const dateB = new Date(dateStringB);
    return dateB.getTime() - dateA.getTime();
}

**Comment by shubham1206agra** [comment]
Added some minor comment in the proposal.

**Comment by quinthar** [comment]
What are the next steps?  Who are we waiting on, and what are waiting on them to do?

**Comment by shubham1206agra** [comment]
Right now, it tries to arrange itself correctly according to the order provided in the description, but due to some bugs and special characters, the order is a little off.

You can check my proposal for the same.

**Comment by shubham1206agra** [comment]
One thing we need to change is the inclusion of RBR in the pinned / GBR report.

**Comment by thebigshotsam** [comment]
## Sorting Logic in Original Code

### Sorting Pinned/GBR Reports and Draft Reports (Alphabetically)

Pinned/GBR reports (`pinnedAndGBRReports`) and Draft reports (`draftReports`) are sorted alphabetically based on the `displayName` property.

```javascript
pinnedAndGBRReports.sort((a, b) => (a?.displayName && b?.displayName ? a.displayName.toLowerCase().localeCompare(b.displayName.toLowerCase()) : 0));
draftReports.sort((a, b) => (a?.displayName && b?.displayName ? a.displayName.toLowerCase().localeCompare(b.displayName.toLowerCase()) : 0));
```
## Sorting Non-archived Reports (Alphabetically or by Most Recent Action)
For non-archived reports (nonArchivedReports), the sorting is based on the display mode. In default mode, the reports are sorted by the combination of the lastVisibleActionCreated date and displayName. In GSD mode, they are sorted alphabetically by displayName.
```javascript

if (isInDefaultMode) {
    nonArchivedReports.sort((a, b) => {
        const compareDates = a?.lastVisibleActionCreated && b?.lastVisibleActionCreated ? compareStringDates(b.lastVisibleActionCreated, a.lastVisibleActionCreated) : 0;
        const compareDisplayNames = a?.displayName && b?.displayName ? a.displayName.toLowerCase().localeCompare(b.displayName.toLowerCase()) : 0;
        return compareDates || compareDisplayNames;
    });
} else {
    nonArchivedReports.sort((a, b) => (a?.displayName && b?.displayName ? a.displayName.toLowerCase().localeCompare(b.displayName.toLowerCase()) : 0));
}
```
## Sorting Archived Reports (Alphabetically or by Most Recent Action)
Similar to non-archived reports, archived reports (archivedReports) are sorted based on the display mode, either by the combination of lastVisibleActionCreated and displayName in default mode or alphabetically by displayName in GSD mode.

```javascript
archivedReports.sort((a, b) => (a?.displayName && b?.displayName ? a.displayName.toLowerCase().localeCompare(b.displayName.toLowerCase()) : 0));
```
## Final Concatenation and Caching
The sorted groups are concatenated into a single array (LHNReports), and this array, along with the cache key, is stored in the cache.

```javascript
const LHNReports = [...pinnedAndGBRReports, ...draftReports, ...nonArchivedReports, ...archivedReports].map((report) => report.reportID);
setWithLimit(reportIDsCache, cachedReportsKey, LHNReports);
```
The sorting involves a combination of alphabetical sorting based on the displayName property and, in some cases, sorting by the `lastVisibleActionCreated` date. The sorting order depends on the display mode (default or GSD).

**Comment by thebigshotsam** [comment]
# Proposal for Sorting Logic Enhancement

## Problem Statement:
The current sorting logic in the Last Heard Navigation (LHN) feature doesn't handle non-alphanumeric characters appropriately when sorting alphabetically. The requirement is to ignore these characters during alphabetical sorting for both "Pinned/GBR Reports" and "Draft Reports."

## Proposed Changes:

### 1. Handling Non-Alphanumeric Characters:
   - Update the sorting logic for `pinnedAndGBRReports` and `draftReports` in the `sortLHNReports` function.
   - Implement a replacement mechanism to remove non-alphanumeric characters before performing alphabetical sorting.

```typescript
pinnedAndGBRReports.sort((a, b) => {
    const cleanA = a?.displayName?.replace(/[^0-9a-zA-Z]/g, '');
    const cleanB = b?.displayName?.replace(/[^0-9a-zA-Z]/g, '');
    return cleanA && cleanB ? cleanA.localeCompare(cleanB) : 0;
});

draftReports.sort((a, b) => {
    const cleanA = a?.displayName?.replace(/[^0-9a-zA-Z]/g, '');
    const cleanB = b?.displayName?.replace(/[^0-9a-zA-Z]/g, '');
    return cleanA && cleanB ? cleanA.localeCompare(cleanB) : 0;
});
```

### 2. Verification of GBR and RBR:
   - Verify that Green/Red Brick Road (GBR/RBR) chats are correctly identified as "pinned" during the sorting process.
   - Ensure that the sort comparator is updated to handle these special cases appropriately.

## Benefits:
- The proposed changes will enhance the alphabetical sorting logic by ignoring non-alphanumeric characters, aligning with the specified requirements.
- GBR/RBR chats will be accurately treated as "pinned" during the sorting process.

## Dependencies:
- Ensure compatibility and minimal impact on existing functionalities.

**Comment by shubham1206agra** [comment]
> > Right now, it tries to arrange itself correctly according to the order provided in the description, but due to some bugs and special characters, the order is a little off.
> 
> Thanks for the feedbacks , So the current bugs are :
> 
>     * RBR are not showing as pinned.
> 
>     * Sorting funtion does not ignore non alphabetical characters.
> 
> 
> > We need to use this inside `getOrderedReportIDs` to get the RBR reports to get them alongside pinned and GBR reports.
> 
> @shubham1206agra Assuming that the code will looks like this code snippet, won't this code introduce any possible performance regression for HT account ?
> 
> ```js
> const hasRoadIndicator = Object.keys(OptionsListUtils.getAllReportErrors(report, ReportActionsUtils.getAllReportActions(report.reportID)) ?? {}).length !== 0;
> 
> if (isPinned || ReportUtils.requiresAttentionFromCurrentUser(report, reportAction) || hasRoadIndicator) {
>     pinnedAndGBRReports.push(report);
> }
> ```

I fear the same actually. But there's no other way to identify RBR in reports as errors is only contained in reportActions.
@quinthar What's your opinion on this? Should we keep RBR reports as pinned or not do this (inclusion of RBR reports may become very slow as we have to iterate on every messages right now to get the RBR status of the report)?

**Comment by quinthar** [comment]
Ah, great question.  However, I'm not sure I understand.  Right now the client already correctly shows the RBR red dot in the LHN, it's just not sorted correctly.  Given that the client is already doing whatever is necessary to get the red dot, why can't we just use that -- why would we need to do some new expensive calculation to determine it?

**Comment by thebigshotsam** [comment]
@quinthar, thanks for raising this point. Let's break down the approach:

### Clarification on RBR Identification:

It's mentioned that the client correctly displays the RBR red dot in the LHN, indicating that the information about RBR status is available client-side. If this is the case, we can leverage this existing client-side information in our server-side logic. The question is valid: why introduce a new expensive calculation if the client is already obtaining the necessary details?

### Proposed Approach:

1. **Review Pinning Logic:**
   - Confirm that the pinning logic correctly identifies and includes RBR reports.
   - Ensure that `report.isPinned` is accurately set for RBR reports, aligning with the client-side representation.

2. **Utilize Client-Side Information:**
   - If the client already fetches and displays the RBR red dot correctly, consider utilizing this information on the server side.
   - Evaluate whether the client-side handling of RBR status can be seamlessly integrated into the server-side logic for sorting.

3. **Address Sorting Issue:**
   - Investigate why RBR reports are not showing as pinned.
   - Ensure that the sorting mechanism aligns with the correct representation of RBR status.
 ```javascript
 // Inside the forEach loop where reports are processed

// ... existing code ...

reportsToDisplay.forEach((report) => {
  // ... existing code ...

  const isPinned = report.isPinned ?? false;
  
  // Check if it's an RBR report and adjust isPinned accordingly
  if (ReportUtils.isRedBrickRoadReport(report)) {
      // Assuming RBR reports should be treated as pinned
      isPinned = true;
  }

  const reportAction = ReportActionsUtils.getReportAction(report.parentReportID ?? '', report.parentReportActionID ?? '');
  if (isPinned || ReportUtils.requiresAttentionFromCurrentUser(report, reportAction)) {
      pinnedAndGBRReports.push(report);
  } else if (report.hasDraft) {
      draftReports.push(report);
  } else if (ReportUtils.isArchivedRoom(report)) {
      archivedReports.push(report);
  } else {
      nonArchivedReports.push(report);
  }
});

// ... existing code ...
```

### Client-Side vs. Server-Side Handling:

Given the existing functionality on the client side, it's worth exploring a solution that optimally integrates with this setup rather than introducing additional complexity. This can potentially lead to a more efficient and streamlined approach.

**Comment by shubham1206agra** [comment]
> Ah, great question. However, I'm not sure I understand. Right now the client already correctly shows the RBR red dot in the LHN, it's just not sorted correctly. Given that the client is already doing whatever is necessary to get the red dot, why can't we just use that -- why would we need to do some new expensive calculation to determine it?

Actually all the RBR calculation are deferred to later steps while rendering reports. I can adjust the steps to make calculation not redundant, but might still be a costly calculation.

Let me check if I can have some optimizations in the logic, and will let you know here.

**Comment by youssef-lr** [comment]
@shubham1206agra I looked into this, and it seems like we already `getAllReportErrors` any the LHN re-renders, if it's not affecting performance now then I think it's fine if we also call it when re-ordering reports, which doesn't happen as often as when the LHN re-renders.

Also, I think we should include checking for violations inside `getAllReportsErorrs`, you can check [here](https://github.com/Expensify/App/blob/963d957f5f8c0fd2ef6acbf6c43a58684e383be3/src/hooks/useViolations.ts) how we extract the violation message. I think we'll probably need a new method to extract the messages as we can't use that hook outside a component.

>I can adjust the steps to make calculation not redundant

If there's a way to make that function somehow only triggered when its dependencies are changed, it would be great, but I'm not sure how feasible is that given that it lives outside components.

What do you think?

**Comment by shubham1206agra** [comment]
@youssef-lr Can you hold this on https://github.com/Expensify/App/pull/31448 in that case?

**Comment by quinthar** [comment]
Looks like https://github.com/Expensify/App/pull/31448 was merged; taking off of hold.  What's the next step here, who's doing it, and when will it be done?

**Comment by quinthar** [comment]
Bump, what's the next step here?

**Comment by shubham1206agra** [comment]
@fedirjh Again here problem would be any new message would trigger the change in whole structure.

**Comment by hannojg** [comment]
Yeah sure, let me read through all it! Give me a moment to write an insight from my perspective

**Comment by lindboe** [comment]
> Also, I think we should include checking for violations inside getAllReportsErorrs, you can check [here](https://github.com/Expensify/App/blob/963d957f5f8c0fd2ef6acbf6c43a58684e383be3/src/hooks/useViolations.ts) how we extract the violation message. I think we'll probably need a new method to extract the messages as we can't use that hook outside a component. Also, only violations with type violation or warning should be taken into account here. (cc @lindboe as I see you're working on client side violations)

@cead22 I think this is something you need to comment on (and cc @cdanwards ). I think the plan we talked about and is in the violations doc is to NOT include violations in `allReportErrors`, but instead always append violations logic as an extra check in-line:

> In src/libs/SidebarUtils.js we need to update [this line](https://github.com/Expensify/App/blob/cfd6f7b5a1/src/libs/SidebarUtils.js#L247) with
`result.brickRoadIndicator = !_.isEmpty(result.allReportErrors) || ReportUtils.transactionThreadHasViolations(report) ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : '';`
That way src/components/LHNOptionsList/OptionRowLHN.js will show the RBR

> In order to show the reports with violations at the top of the LHN, In src/libs/SidebarUtils.js
We’ll update [this](https://github.com/Expensify/App/blob/dee9b3110a7ebe0aa250c3f966d6a69925d916f4/src/libs/SidebarUtils.ts#L363) to return true if `Object.keys(result.allReportErrors ?? {}).length !== 0 || ReportUtils.transactionThreadHasViolations(report)`

I'm not very familiar with the problem this ticket describes; are we going to run into problems with a conflicting approach here @youssef-lr ?

**Comment by hannojg** [comment]
>  We can get an ids array of reports that have errors …

Minor thing, but if we are already iterating over that array, let's use an object structure. This time the lookup time for a report action later in `getOrderedReportsID` will be just O(1)

> … then pass the array to the getOrderedReportIDs

Sounds good to me!

> we should check what are the necessary props for detecting the errors [and add that to the selector]

Yes, agreed.

As you @fedirjh pointed out, by following the existing scheme of implementation in `SidebarLinksData` we should be guarded against performance pitfalls 😊

**Comment by youssef-lr** [comment]
@fedirjh nice yeah, I was thinking about memoizing report errors as well. I think we can take that a step further and memoize it in `OptionRowLHNData` with these dependencies (if eslint won't complain) `[report.errors, report.errorFields, reportActions]` as for the report we only care about those 2 props to decide if it has an error. This way the function (which iterates over _every_ report action will not be triggered if a random report prop changes).  We can pass the memozied `reportErrors` keyed by reportID to `SidebarUtils.getOptionData` [here](https://github.com/Expensify/App/blob/b74be723548d45d47dde7fe29aff5eafe16b290f/src/components/LHNOptionsList/OptionRowLHNData.js#L90).

What do you think @fedirjh @hannojg?

**Comment by cead22** [comment]
> I think the plan we talked about, and are implementing in #31411, and is in the violations design doc, is to NOT include violations in `allReportErrors`, but instead always append violations logic as an extra check in-line:

@lindboe I think you can proceed as planned there, and the refactor to put violations in `allReportErrors` as part of this issue seems like it should work fine

**Comment by luacmartins** [comment]
I can help move this forward. @fedirjh would you mind giving us a summary of the current proposal we're exploring and the many optimizations we mentioned in this thread?

**Comment by luacmartins** [comment]
@fedirjh thanks for the detailed analysis and highlighting the performance concerns. 

Given that we're in agreement on @shubham1206agra's proposal, maybe the next step would be for @shubham1206agra to create a POC and time it on a HT account. That will help us confirm that we will indeed have performance degradation and how much of an issue that'll be. From there, we can explore solutions to mitigate any performance issues.

@hannojg would the reassure performance tests be a good way to measure the impact on SidebarLinksData render time?

**Comment by mountiny** [comment]
> @hannojg would the reassure performance tests be a good way to measure the impact on SidebarLinksData render time?

We have a ressure tests for LHN but I dont think they do specifically mock pinned and GBR/RBR chats https://github.com/Expensify/App/blob/6cfdd6f47fe5f932b70f440104accffea586ef60/tests/perf-test/SidebarLinks.perf-test.js#L63-L77

@OlimpiaZurek Do you think we could add a scenario for this case?

**Comment by luacmartins** [comment]
Hmm I think at this point we just need to measure the performance rendering the LHN with the new logic. Asserting the correct order would be done in a separate test I imagine.

**Comment by hannojg** [comment]
I just checked the reassure tests and it seems that they create 10k mock action items, which seems to be a good setup to test the performance. In addition to that it should be covered by the e2e performance regression tests. So I think in terms of tests we should be on a safe ground.

**Comment by luacmartins** [comment]
Sounds good, so I think we can go ahead with a POC and see what the tests reveal.

**Comment by mountiny** [comment]
> Hmm I think at this point we just need to measure the performance rendering the LHN with the new logic. Asserting the correct order would be done in a separate test I imagine.

Fair, what I meant is that the mocks probably do not add any variation to the type of reports which could make the sorting faster than real usecase with different report types, archived chats, GBR, hidden reports etc

**Comment by luacmartins** [comment]
Yea, I'm sure the real use case will be different and most likely slower. I think we start simple with the existing tests and then move on to more complex cases.

**Comment by OlimpiaZurek** [comment]
> > @hannojg would the reassure performance tests be a good way to measure the impact on SidebarLinksData render time?
> 
> We have a ressure tests for LHN but I dont think they do specifically mock pinned and GBR/RBR chats
> 
> https://github.com/Expensify/App/blob/6cfdd6f47fe5f932b70f440104accffea586ef60/tests/perf-test/SidebarLinks.perf-test.js#L63-L77
> 
> @OlimpiaZurek Do you think we could add a scenario for this case?

I don't think a new scenario is needed here. I think we just need to update the mocks we passed to the SidebarLinks to include pinned and GBR/RBR chats.

**Comment by luacmartins** [comment]
@shubham1206agra would you be available to work on a POC of your solution?

**Comment by shubham1206agra** [comment]
I will be available from 30th January. I am OOO right now.

**Comment by youssef-lr** [comment]
>>I think we can take that a step further and memoize it in OptionRowLHNData with these dependencies [report.errors, report.errorFields, reportActions]

>@youssef-lr Why do we need to do that? The report ordering logic is done in SidebarLinksData which has a higher hierarchy order than OptionRowLHNData.

@fedirjh Sorry what I'm suggesting there is related the LHN re-rendering in general, every time the LHN re-renders we're iterating over all report actions even if there was no changes to them. So in addition to what you suggested, I think we should memoize each report's errors in `OptionRowLHNData` as well. The reason for memoizing the errors per report/reportActions (instead of all reports and reportActions) is that we will be re-running the function only for a single report/reportActions when this report has any changes.

If we memozie all of the report actions, then we could re-run `getAllReportErrors` for `Report B` even though it was just `Report B` that got changed. Again this is related to the LHN only, and not the sorting function called in `SidebarLinksData`, I think it's related because when we re-sort reports we're still causing a re-render in the LHN. What do you think?

**Comment by youssef-lr** [comment]
@fedirjh In my very basic test using a `console.log` from inside `getAllReportErrors`, I could see it being called unnecessarily, even when I'm just resizing the window.

https://github.com/Expensify/App/assets/9680864/e3239895-5f15-487e-8e58-04367c61c2c6

**Comment by youssef-lr** [comment]
Also, one more optimization is making the `useMemo` dependencies the following: `[report.errors, report.errorFields, reportActions]`, I don't see that the other report props we're accessing inside `getAllReportErrors` are subject to change (`reportID, parentReportID, parentReportActionID`). Eslint will complain though but I think it would make sense to silence it in this case.

**Comment by luacmartins** [comment]
@shubham1206agra hope you enjoyed your ooo time! Are you back and available to work on a POC for this issue?

**Comment by shubham1206agra** [comment]
Yes. I will start working on it.

**Comment by quinthar** [comment]
@shubham1206agra can you give an ETA for how long this will take?

**Comment by shubham1206agra** [comment]
Is Monday fine with you?

**Comment by luacmartins** [comment]
That seems reasonable given the complexity here. Looking forward to the POC

**Comment by shubham1206agra** [comment]
@luacmartins @fedirjh Please check this draft PR https://github.com/Expensify/App/pull/35817.

**Comment by luacmartins** [comment]
We have a draft PR.

**Comment by JmillsExpensify** [comment]
Draft PR still linked above.

**Comment by luacmartins** [comment]
@youssef-lr will be taking this one while I'm ooo next week!

**Comment by quinthar** [comment]
what exactly remains?  Who is doing it, and when?

**Comment by youssef-lr** [comment]
We have this PR from @shubham1206agra but it's not complete yet, it's missing a few performance optimizations we discussed in this issue as well as sorting reports alphanumerically. 

@shubham1206agra can you please make sure your PR covers the following:

1. Wrap report errors in `useMemo` as described [here](https://github.com/Expensify/App/issues/33029#issuecomment-1904467031) and pass them to `getOrderedReportIDs`, we should not be making a call to `getAllReportErrors` from that lib as it will affect performance.
2. Wrap report errors in `useMemo` when rendering the LHN as described [here](https://github.com/Expensify/App/issues/33029#issuecomment-1915313669), this is a general LHN optimization and not necessarily related to this issue.
3. Add support for displaying an RBR next to reports that have a violation of type `violation`.
4. Use `reportErrors` & `transactionViolations` to decide if a report has an RBR, and pin it accordingly in `getOrderedReportIDs`. For violations, we seem to already have a handy variable for this `doesReportHaveViolations`.

**Comment by JmillsExpensify** [comment]
Looks like C+ also has questions for the contributor waiting in the PR. @shubham1206agra are you able to prioritize this in the coming days, so we can get this PR merged?

**Comment by youssef-lr** [comment]
Not overdue, we're still waiting for the comments to be addressed in the PR.

**Comment by youssef-lr** [comment]
@shubham1206agra I'm not sure how that PR interferes with this one?

**Comment by youssef-lr** [comment]
That PR has 16 files changed, putting it on HOLD will most likely cause it to have many conflicts every day. I personally don't see any reason why we should wait for that PR, I really don't see how it conflicts with this one, do you mind expanding on why you think we need to wait for that PR?

**Comment by youssef-lr** [comment]
@fedirjh has ran some tests and the results look promising, it looks like there is no performance hit with adding report errors to the sorting logic.

**Comment by youssef-lr** [comment]
We can start testing this @fedirjh.

**Comment by youssef-lr** [comment]
Reviewing PR now.

**Comment by youssef-lr** [comment]
Bummer. How about we make pinning RBR reports only rely on the transaction violations coming from the violations onyx key? I think the performance hit is coming from the fact that we iterate over all report actions to find out if they have an error. The errors stored in the report actions are just API errors, which I think are rare.

**Comment by JmillsExpensify** [comment]
Removing the payment from title as we still work on the implementation.

**Comment by youssef-lr** [comment]
Yes I think we should. I'll get back to you this week @shubham1206agra as I'm busying finalizing a new feature.

**Comment by luacmartins** [comment]
What's the latest here? I think we merged a fix for this issue and it's done? Are we missing payments?

**Comment by youssef-lr** [comment]
@luacmartins we need to redo the fix, it caused performance issues and we reverted it.

**Comment by youssef-lr** [comment]
I think let's redo it and see if reassure-perf complains

**Comment by JmillsExpensify** [comment]
Is this being actively worked on? If not, my preference is to close.

**Comment by luacmartins** [comment]
I don't think so, but @youssef-lr would be able to confirm

**Comment by JmillsExpensify** [comment]
Youssef please open back up if it's being worked on. Otherwise, I think we have more important priorities to focus on.
